### PR TITLE
preview deployment

### PR DESCRIPTION
**Pull request by rajatjindal**
this PR adds `fermyon/actions/spin/preview` action. It can do following:

### New PR open

- build and deploy preview to Fermyon Cloud as `<original-name>-pr-<pr-number>`
- update comment on the PR with preview link
- the deployment will fail if the number of apps limit has exceeded on Fermyon Cloud

### New code pushed to same PR

- build and deploy updated preview
- it will use the same app name that was used for deployment when PR was originally opened

### PR closed

- removes the PR deployment from Fermyon Cloud

### CODE REVIEW

**Comment by rajatjindal** [comment]
this PR adds `fermyon/actions/spin/preview` action. It can do following:

### New PR open

- build and deploy preview to Fermyon Cloud as `<original-name>-pr-<pr-number>`
- update comment on the PR with preview link
- the deployment will fail if the number of apps limit has exceeded on Fermyon Cloud

### New code pushed to same PR

- build and deploy updated preview
- it will use the same app name that was used for deployment when PR was originally opened

### PR closed

- removes the PR deployment from Fermyon Cloud

File: README.md [hunk 1]
```diff
@@ -179,3 +180,49 @@ jobs:
           manifest_file: example-app/spin.toml
           fermyon_token: ${{ secrets.FERMYON_CLOUD_TOKEN }}
 ```
+
+
+## `fermyon/action/spin/preview`
```
Comment (vdice) [inline README.md:185]:
```suggestion
## `fermyon/actions/spin/preview`
```

(The other headers may need updating `action` -> `actions` as well; can be a follow-up of course)

File: README.md [hunk 2]
```diff
@@ -7,6 +7,7 @@ It enables the following usecases:
 - setup spin and plugins (`fermyon/actions/spin/setup`)
 - build and push to OCI registry (`fermyon/actions/spin/push`)
 - deploy to Fermyon Cloud (`fermyon/actions/spin/deploy`)
+- deploy PR preview to Fermyon Cloud (`fermyon/actions/spin/preview`)
```
Comment (vdice) [inline README.md:10]: What do you think about having `fermyon/actions/spin/preview` link to the corresponding section below?  (Could be done with all of these, so eligible to be a follow-up.)

File: README.md [hunk 3]
```diff
@@ -179,3 +180,49 @@ jobs:
...
+## `fermyon/action/spin/preview`
+
+Build and deploy the `spin` app preview to Fermyon Cloud
+
+### Inputs
+
+| Name          | Required | Description                                                                       | Default   |
+| ------------- | -------- | --------------------------------------------------------------------------------- | --------- |
+| fermyon_token | Required | Fermyon Cloud Personal access token for deploying the `spin` app to Fermyon Cloud | -         |
+| manifest_file | Optional | Path to `spin.toml`. Used with the `build`/`deploy` command.                      | spin.toml |
```
Comment (vdice) [inline README.md:194]: Should we update "Used with the `build`/`deploy` command." to include `preview`?  Or maybe remove this sentence?

File: src/actions.ts [hunk 4]
```diff
@@ -51,4 +55,44 @@ export async function registryLogin(): Promise<void> {
...
+    const manifestFile = getManifestFile()
+    const spinConfig = spin.getAppManifest(manifestFile)
+
+    const realAppName = spinConfig.name
+    const previewAppName = `${realAppName}-pr-${prNumber}`
+
+    core.info(`🚀 deploying preview as ${previewAppName}`)
+    const metadata = await cloud.deployAs(previewAppName, manifestFile)
+
+    const comment = `🚀 preview deployment successful and available at ${metadata.base}`
```
Comment (vdice) [inline src/actions.ts]: Wondering if we want to add 'Spin' and/or 'Fermyon Cloud' in the comment?

File: README.md [hunk 5]
```diff
@@ -179,3 +180,49 @@ jobs:
...
+          version: canary
+          plugins: js2wasm
+
+      - name: build and deploy
+        uses: fermyon/actions/spin/preview@v1
+        with:
+          manifest_file: example-app/spin.toml
+          fermyon_token: ${{ secrets.FERMYON_CLOUD_TOKEN }}
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          undeploy: ${{ github.event.pull_request && github.event.action == 'closed' }}
```
Comment (vdice) [inline README.md:227]: I wonder if we want to explore having this same condition (i.e. PR closed) be the default in the action code?

File: src/cloud.ts [hunk 6]
```diff
@@ -29,6 +41,81 @@ export class Route {
...
+            throw `no app found with name ${name}`
+        }
+
+        return app.id;
+    }
+
+    async deleteAppById(id: string): Promise<void> {
+        const resp = await this._httpclient.get(`${this.base}/api/apps`)
+        if (resp.message.statusCode !== httpm.HttpCodes.OK) {
+            throw `expexted code ${httpm.HttpCodes.OK}, got ${resp.message.statusCode}`
```
Comment (vdice) [inline src/cloud.ts]:
```suggestion
            throw `expected code ${httpm.HttpCodes.OK}, got ${resp.message.statusCode}`
```

**Review by vdice — COMMENTED** [review]
This functionality is really exciting!  All looking good to me, just some comments/Qs.

File: README.md (hunk 3 above)
Comment (vdice) [reply README.md:194]: Oh, n/m, I see how this applies to `spin build` and `spin deploy`.  Disregard :)

File: README.md (hunk 1 above)
Comment (rajatjindal) [reply README.md:185]: done

File: src/actions.ts (hunk 4 above)
Comment (rajatjindal) [reply src/actions.ts]: changed it to `preview deployed successfully to Fermyon Cloud and available at`

File: README.md (hunk 5 above)
Comment (rajatjindal) [reply README.md:227]: In event section of workflow we need 'closed' type to trigger this behavior. and I think having this in workflow also serves as documenting when this preview will be undeployed

not very strongly against having the default, so if you insist, I will add this as default :)

File: src/cloud.ts (hunk 6 above)
Comment (rajatjindal) [reply src/cloud.ts]: done :)

File: README.md (hunk 2 above)
Comment (rajatjindal) [reply README.md:10]: @macolso has opened a PR to update the README, so we will get this addressed there.

File: README.md (hunk 5 above)
Comment (vdice) [reply README.md:227]: Sounds good, let's go with this for now and we can always revisit.

File: src/actions.ts (hunk 4 above)
Comment (vdice) [reply src/actions.ts]: I'm not seeing that change included in the latest, but we can always adjust/update later.

File: src/cloud.ts (hunk 6 above)
Comment (vdice) [reply src/cloud.ts]: Oh I think you might not have pushed your latest changes -- can you double-check?

File: README.md (hunk 5 above)
Comment (rajatjindal) [reply README.md:227]: thank you Vaughn

File: src/actions.ts (hunk 4 above)
Comment (rajatjindal) [reply src/actions.ts]: i forgot to push the code :), pushed now.

File: src/cloud.ts (hunk 6 above)
Comment (rajatjindal) [reply src/cloud.ts]: yep, forgot to push the code. done now :)

**Review by vdice — APPROVED** [review]
LGTM!

**Pull request merged**